[MenuUnstyled] Create MenuUnstyled and useMenu

michaldudak opened the pull request:
This PR introduces the unstyled menu component and friends: `MenuUnstyled`, `MenuItemUnstyled`, `useMenu` and `useMenuItem`.
This is a rewrite of the components used in Material UI and it's based on the new `useListbox` hook.

The unstyled Menu differs from its Material counterpart as it contains only the list of items, without the element that triggers the appearance of the list. It's more similar to Material UI's MenuList in this matter. This helps developers make it more customizable. An example of creating a button triggering a menu is included in the docs.

This PR does not enable support for nested menus. This is to be implemented in a separate PR, taking into account work done in #20591 (cc @EsoterikStare).

Preview - https://deploy-preview-30961--material-ui.netlify.app/base/react-menu/

mnajdova commented:
Shouldn't we skip disabled items when using the keyboard? I don't think this is a valid state:
![image](https://user-images.githubusercontent.com/4512430/154973293-0c4316a6-eb05-41a6-812b-3f87e0391f78.png)

mnajdova commented on docs/data/base/components/menu/menu.md:
```diff
@@ -0,0 +1,79 @@
+
+#### CSS classes
+
+The MenuUnstyled does not receive any state classes.
```
```suggestion
The MenuUnstyled does not set any state classes.
```

The components sets these classes no?

eps1lon commented on docs/data/base/components/menu/MenuPopup.js:
```diff
@@ -0,0 +1,169 @@
+    border-bottom: none;
+  }
+
+  &:focus-visible {
```
Note that only the [technology preview of Safari supports this at the moment](https://caniuse.com/?search=%3Afocus-visible). We should recommend the polyfill in the meantime.

Generally, the demos should have the same support matrix the library has in my opinion.

michaldudak replied:
:+1: I forgot to implement the focus-visible polyfill for menu items. Will fix.

eps1lon commented on docs/data/base/components/menu/MenuPopup.js:
```diff
@@ -0,0 +1,169 @@
+      </TriggerButton>
+      <Popper
+        placement="bottom-start"
+        open={isOpen && buttonRef.current != null}
```
This is reading a ref during render which is generally not safe. This is why we have the pattern for `anchorEl` that puts the element in React state instead of using a ref.

michaldudak replied:
Good point, thanks!

eps1lon commented on docs/data/base/components/menu/MenuPopup.js:
```diff
@@ -0,0 +1,169 @@
+    if (isOpen) {
+      contentRef.current?.focus();
+    }
+  }, [isOpen, contentRef]);
```
This may not be doing what you think it does. `contentRef` has no impact on the effect behavior while the effect is **not** synced with `contentRef.current`

michaldudak replied:
This was exactly my intention. `contentRef` isn't really needed in the dependency array.

eps1lon commented:
It's a bit odd to me that the default pattern for a Menu does not use it as a popup. In my perception Menus are always in a popup.

michaldudak reviewed (commented):
@danilo-leal thanks for making it prettier! :)

> Shouldn't we skip disabled items when using the keyboard?

According to https://www.w3.org/TR/wai-aria-practices-1.2/#h-note-17 disabled items should receive focus. The implementations in the wild seem to use both approaches, though. For example, disabled items in menus on Windows do receive focus, while on Macs they don't.

> It's a bit odd to me that the default pattern for a Menu does not use it as a popup. In my perception Menus are always in a popup.

Thanks for the feedback. On the one hand, not having the popup would make it easier for developers to customize the menu and use whatever popup solution they want. On the other hand, though, it makes the most common use case harder to achieve. @mnajdova do you have an opinion on this?

mnajdova commented:
> Thanks for the feedback. On the one hand, not having the popup would make it easier for developers to customize the menu and use whatever popup solution they want. On the other hand, though, it makes the most common use case harder to achieve. @mnajdova do you have an opinion on this?

I must admit I had a similar thought. When I first saw it, my thought was, isn't this the same as the list. But the keyboard navigation is different, so this was the justification I had at that moment. My initial assumption was that the `MenuUnstyled` would be inside a popup and would even accept an open prop.

Let's benchmark how it is implemented in other libraries:

Reakit - https://reakit.io/docs/menu/#main
Headless UI - https://headlessui.dev/react/menu
Mantine - https://mantine.dev/core/menu/

Looks like the logic is somehow provided from the library.

siriwatknp commented:
I am thinking about using the `useMenuItem` hook inside the Joy's `ListItemButton`.

From the design perspective, I think the content in the MenuPopup is a bunch of list components with keyboard support. So I want to leverage the list components without providing MenuItem.

This is what I thought about using the Menu in Joy.

```js
import Menu from '@mui/joy/Menu';
import List from '@mui/joy/List';
import ListItem from '@mui/joy/ListItem';
import ListItemButton from '@mui/joy/ListItemButton';

<Menu>
  <List sx={{ ...CSS variables of the List }}>
    <ListItem>
      <ListItemButton>Option 1</ListItemButton>
    </ListItem>
    <ListItem>
      <ListItemButton>Option 2</ListItemButton>
    </ListItem>
    ...
  </List>
</Menu>
```

Benefits of this approach:
- `<List>...</List>` can be moved in/out of Menu without any changes 😲
- Developer does not need to learn a new API (MenuItem)
- We need to create CSS variables is only for the List (less component to maintain)

**How to do this?**
I am thinking about using `useMenuItem` inside Joy's `ListItemButton`. If there is a Menu context, spread the properties, otherwise don't use it.

```js
// ListItemButton in Joy
function ListItemButton() {
  const menuContext = useMenuItem({ component: 'button', ref }) // what's component for?
  return (
    <ListItemButtonRoot {...menuContext?.getRootProps()} />
  )
}
```

Is this possible? @michaldudak

michaldudak commented:
> I am thinking about using the useMenuItem hook inside the Joy's ListItemButton.

@siriwatknp Sure! Hooks are meant to be composed to make higher-level components. In case of `useMenuItem` it's just a small utility to register the item within the `Menu` to be able to be notified about whether it should be selected.

> From the design perspective, I think the content in the MenuPopup is a bunch of list components with keyboard support

Indeed! Just note that the menu implements a roving focus (only the currently selected item has `tabIndex: 0` and the rest has `-1`, so it's not possible to navigate through items with the <kbd>Tab</kbd> key - not sure if this is something you want.

> `useMenuItem({ component: 'button', ref }) // what's component for?`

The `useButton` hook (which is used by `useMenuItem` needs to know if it's applied on a `button` element or something else. Props such as `role`, `disabled`/`aria-disabled`, and event handlers depend on it.
I am considering changing this API as I don't really like how it's used. I will refactor the Joy implementation if I do so.

mnajdova commented on docs/data/base/components/menu/MenuSimple.js:
```diff
@@ -0,0 +1,164 @@
+  };
+
+  const handleButtonKeyDown = (event) => {
+    if (event.key === 'ArrowDown' || event.key === 'ArrowUp') {
```
If the `ArrowUp` is pressed, the last element should be foced when the menu opens.

michaldudak commented:
> esc should close the menu and focus the trigger element (implemented in Material-UI)

Implemented in the demos. I'll consider creating an unstyled MenuButton (or DropdownButton) that has this functionality built-in. It could also be used for a Select.

> you should be able to select the element by typing the first letter (implemented in Material-UI)

This makes more sense in cases where items are ordered alphabetically (which is rarely the case in a menu). I am going to implement it anyway in Listbox (which is what Select and Menu use internally).

> the button opening the Menu should have the corresponding aria attributes: aria-haspopup, aria-controls (to reference the menu it opens), aria-expanded (when the menu is opened) (implemented in the demos for Material-UI)

This is now fixed

> the button opening the Menu should open the menu on Down/Up arrow (not implemented in Material-UI)
> If the ArrowUp is pressed, the last element should be foced when the menu opens.

:+1: Working on it

michaldudak commented:
> the button opening the Menu should open the menu on Down/Up arrow (not implemented in Material-UI)
If the ArrowUp is pressed, the last element should be foced when the menu opens.

@mnajdova this should work as expected now.

mnajdova commented on docs/data/base/components/menu/MenuSimple.js:
```diff
@@ -0,0 +1,181 @@
+      event.preventDefault();
+      setAnchorEl(event.currentTarget);
+      if (event.key === 'ArrowUp') {
+        menuActions.current?.highlightLastItem();
```
I don't like that we need to use `useImperativeHandle` here. Is there a different option? One additional reason to have the MenuButton as a standalone component.

mnajdova replied:
In addition to this:

```
      <TriggerButton
        type="button"
        onClick={handleButtonClick}
        onKeyDown={handleButtonKeyDown}
        ref={buttonRef}
        aria-controls={isOpen ? 'simple-menu' : undefined}
        aria-expanded={isOpen || undefined}
        aria-haspopup="menu"
      >
        Language
      </TriggerButton>
```

Has too much boilerplate which is important to not be missed. I would propose to tackle this in the next PR.

michaldudak replied:
Yes, I absolutely agree it's too much code. I will create the MenuButton for sure.
As for `useImperativeHandle` - I realize it's not quite in the spirit of React, but I couldn't think of a better option that would work with the current MenuUnstyled design. I'm happy to discuss alternatives.

mnajdova commented on packages/mui-base/src/MenuUnstyled/MenuUnstyled.tsx:
```diff
@@ -0,0 +1,186 @@
+  };
+
+  return (
+    <Popper {...popperProps} ref={forwardedRef}>
+      <Listbox {...listboxProps}>
```
Just to confirm, we talked about merging these one into one DOM element right?

mnajdova reviewed (approved):
The behavior looks good. Next steps would be:
- create `MenuButton` component
- merge the two dom elements we have in the `MenuUnstyled` - root and listbox into one.
